### ISSUE TITLE
Bug(arrow-row): calling `convert_raw` function cause "offset overflow" panic

### ISSUE THREAD

**Issue by JasonLi-cn** [apache/arrow-rs#6112]
**Describe the bug**

- Datafusion Table Info: 
Having a `http_url` column, which DataType is `Utf8`, and it has a lot of distinct values.

- Datafusion SQL

```sql
select http_url from tab group http_url
```

- Panic Info
<img width="1025" alt="image" src="https://github.com/user-attachments/assets/76c1d0e4-b037-4444-aa40-b1fdc00d4466">

- Reason
`GroupValuesRows` in Datafusion stores rows using `Rows`, and this bug may be triggered when calling `emit` function.
https://github.com/apache/arrow-rs/blob/49e714de6e951169d0d5e73381af247ad0230fcf/arrow-row/src/variable.rs#L217-L226

**To Reproduce**

**Expected behavior**

**Additional context**

**Comment by alamb** [comment]
In newer versions of DataFusion I would expect that grouping on a string column would not use row format, but instead usse the special GroupValuesBytes:

https://github.com/apache/datafusion/blob/7db4213b71ed9e914c5a4f16954abfa20b091ae3/datafusion/physical-plan/src/aggregates/group_values/bytes.rs#L27

**Comment by alamb** [comment]
The bug you describe certainly can happen if there are large numbers of distinct large strings 🤔

**Comment by tustvold** [comment]
FWIW in general offset overflows do yield panics in arrow-rs, the additional plumbing for error handling what is almost always an unrecoverable error has been hard to justify

**Comment by JasonLi-cn** [comment]
> The bug you describe certainly can happen if there are large numbers of distinct large strings in a multi-column group 🤔

Yes, this problem was first discovered in the case of a group by multi-column.

**Comment by JasonLi-cn** [comment]
> FWIW in general offset overflows do yield panics in arrow-rs, the additional plumbing for error handling what is almost always an unrecoverable error has been hard to justify, although I suspect in this case it could be made into an `ArrowError` without it being a breaking change.
> 
> Edit: I've updated this to be an enhancement, panics are not a bug

Do you mean we need to add a new function like the following?
```rust
pub fn try_decode_binary<I: OffsetSizeTrait>(
    rows: &mut [&[u8]],
    options: SortOptions,
) -> Result<GenericBinaryArray<I>, ArrowError> {
    ...
}
```

**Comment by alamb** [comment]
> Do you mean we need to add a new function like the following?

Maybe you could add a check in your code (or in datafusion 🤔 ) on the size of the string buffer and make a new record batch if they exceed 2GB or something. This might be related: https://github.com/apache/datafusion/issues/9562


Making a single array with more than 2GB of string data is likely to be non ideal in a bunch of ways
